Bayes

Bob-IT opened the pull request:
This Branch creates a dialogue to view and delete Bayesian and Non Bayesian information.
As part of this I needed to change qof function wrap_gvalue_function to test for a KvpFrame as I was getting the following errors in the trace file.
* 10:16:15  WARN <qof.kvp> [gvalue_from_kvp_value()] Error! Attempt to transfer KvpFrame!
* 10:16:15  WARN <qof.kvp> [gvalue_from_kvp_value()] Error! Invalid KVP Transfer Request!

jralls commented:
I blocked passing KvpFrames in GValues *explicitly* to prevent direct manipulation of Kvp outside of the engine. The Kvp handling *must* be done in src/engine/Account.c unless you want to abstract it into a class (yes, write it in C++11) that loads an account's map from Kvp and manipulates it in the class. Using the backends for that purpose is OK, but note that I'm working on converting them to C++ right now, so you might want to wait a few days for me to get that finished and merged if you decide to pursue that.

Since this is new code it should comply as much as possible to the [Roadmap](http://wiki.gnucash.org/wiki/Roadmap), in particular the MVC goal. We want only code that is directly required by Gtk in the gnome directories so that we can more easily implement other GUIs -- even a Gtk3 migration will be made much easier with this separation.

jralls commented:
More generally you could help a lot by extracting all of the copied-and-pasted code from your reg2 project and its src/register counterparts, rewriting it in nice clean C++11, and putting it in a new src/controllers directory.

Oh, and welcome to Github! I hope you'll find this environment more comfortable for large changes than bug reports. Remember, you can fix pull requests with new commits on your feature branch. You can even force-push rebased changes onto the same branch. You only need to make a new PR if you have to nuke a branch and start over.

Bob-IT commented:
I have moved the KVP parts to Account.c with two main functions returning a GList of structure kvp_info. Also added a get and delete kvp info based on path for an account to make it easier to process. Hopefully this makes it more MVC compliant.
There is one further thing I would like to do in this area, that is when a Bayesian entry is saved the full account path is used as a key but this is lost when the separator or account name is changed.  What I was thinking that this full name be replaced with a string of the account guid, this keeps the existing kvp layout. Old entries can be converted when used but not sure if it would be better to change them all at once. Is there a mechanism to do this ? 

With regards to c++, I thought I would get it working and then see if I can convert it.

jralls commented on src/engine/Account.c:
```diff
@@ -5484,6 +5484,211 @@ gnc_account_imap_add_account_bayes (GncImportMatchMap *imap,
+    PINFO("build_bayes_layer_two: account '%s', token_count: '%ld'",
+                                  (char*)key, (long)g_value_get_int64(value));
+
+    probability = g_strdup_printf ("%ld", g_value_get_int64 (value));
```
A gint64 is a long only in some 64-bit builds; on 32-bit builds it's a long long. GLib provides G_GINT64_FORMAT which always expands to the right value.

jralls commented:
That's better, though I think some more abstraction would be better still.
Calling it the "Bayesian Editor" is wrong, though, because it's able to edit other things besides the Bayesian matches. "Import Map Editor" maybe, with "imap" instead of "bayes" for the function and widget names?

"Probablity" isn't right, either. The score isn't a probability, it's a count of times the token has been associated with the mapped account. The probability comes from adding up all scores for different accounts that use that token to make a fraction then averaging the fractions for all of the tokens in a description string for each mapped account found.

I think that a flat list of matches is a bit overwhelming. My database has over 15000 token entries. How about a Tree View sorted by Account, Mapped Account, and token/description list? I realize that that makes it a bit more cumbersome to code the Online list, but I think it will be a lot easier on the user.

gnc_account_delete_kvp() and gnc_account_get_kvp_text() leak too much implementation detail. Make them gnc_account_get_map_entry(), gnc_account_delete_map_entry(), etc. The idea is to conceal the very existence of KVP from outside of the engine.

Separate get_account_info() into get_account_info_bayes(), get_account_info_nbayes(), and get_account_info_online(), each with it's own iteration through the account list. Testing the type every pass through the loop is dumb even though the optimizer should be able to pull it out, and it also makes the function too long and harder to grok. gnc_account_for_each_descendant() would be a better choice than rolling your own for loop as it makes the dialog less dependent upon Account's internal representation.

jralls commented:
Yes, the Bayes map should use mapped account GUID like the NBayes matcher does. A scrub function that runs at load time is fine.

Bob-IT commented:
I think I have made the recommended changes and seems to work quite well. I have changed from a list store to a tree store and added a filter option. If all OK I will make the changes to use the GUID for Bayes, not sure if I can continue making these changes on the same branch.

jralls commented on src/gnome/dialog-imap-editor.c:
```diff
@@ -0,0 +1,759 @@
+    for (row = g_list_first (list); row; row = g_list_next (row))
+    {
+	if (gtk_tree_model_get_iter (model, &iter, row->data))
+        {
```
If you're writing a big nested scope with its own variables like this it means that it needs to be a separate function.

jralls commented on src/gnome/dialog-imap-editor.c:
```diff
@@ -0,0 +1,759 @@
+
+    while (valid)
+    {
+        GtkTreePath *tree_path;
```
See above about nested blocks.

jralls commented on src/gnome/dialog-imap-editor.c:
```diff
@@ -0,0 +1,759 @@
+        for (node = imap_list;  node; node = g_list_next (node))
+        {
+            struct imap_info *imapInfo;
+
```
Declare and initialize in one statement if you can. It reduces the risk of having a path where you didn't initialize.

jralls commented on src/libqof/qof/qofinstance.cpp:
```diff
@@ -1292,8 +1292,16 @@ struct wrap_param
 static void
 wrap_gvalue_function (const char* key, KvpValue *val, gpointer data)
 {
```
This shouldn't be necessary any more.

jralls commented:
The tree view is definitely easier on the eyes and makes it clearer that the matching is per-account. The filter function helps a lot too. I think that most users are going to be confused by Bayesian keys, but the worst that can happen is that they have to retrain a lot after deleting the wrong keys.

I'll merge this after you address the style comments.

Bob-IT commented:
Added some functions to simplify delete and filter parts.
I tried removing the qof changes but ended up with the errors I had before. I think it would work if I started one level down in the kvp tree but for bayes I am starting at 'import-map-bayes' which is followed by another frame and then values as below.

      <slot:key>import-map-bayes</slot:key>
      <slot:value type="frame">
        <slot>
          <slot:key>NEW</slot:key>
          <slot:value type="frame">
            <slot>
              <slot:key>Bank Charges</slot:key>
              <slot:value type="integer">2</slot:value>
            </slot>
          </slot:value>
          <slot:key>'FRNDS</slot:key>
          <slot:value type="frame">
            <slot>
              <slot:key>Clothing</slot:key>
              <slot:value type="integer">7</slot:value>
            </slot>
            <slot>
              <slot:key>Healthcare</slot:key>
              <slot:value type="integer">29</slot:value>
            </slot>
          </slot:value>
        </slot>
      </slot:value>

jralls commented:
Ah, I see.  Trapped by design and the mix of C and C++.
The Right Solution is to abstract further, creating in a C++ class ImportMapBayes that hides the KVP storage and presents a cleaner C API. That would allow cleaning up import-backend.c too.

Feel like taking that on? If not, I can merge this and do the abstraction later when I convert Account.

Bob-IT commented:
Can you merge this and I will make a new branch to change the Bayes to use the GUID and that will complete this set of changes. I can then have a look at converting Imap to c++ as a way to learn.

jralls commented:
OK, merged.
